subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildSessionScopeServices.java

```diff
@@ -165,7 +165,11 @@ ScriptSourceHasher createScriptSourceHasher(FileHasher fileHasher) {
     }
 
     FileSystemSnapshotter createFileSystemSnapshotter(FileHasher hasher, StringInterner stringInterner, FileSystem fileSystem, FileSystemMirror fileSystemMirror) {
-        return new DefaultFileSystemSnapshotter(hasher, stringInterner, fileSystem, fileSystemMirror, DirectoryScanner.getDefaultExcludes());
+        // added to fix  MD5 error when .gradle project cache is part of the root source code dirs (typically in native builds)
+        // issue: https://github.com/gradle/gradle/issues/5941
+        DirectoryScanner.addDefaultExclude("**/.gradle/**"); // should never include the .gradle directory in a snapshot
+        FileSystemSnapshotter retval = new DefaultFileSystemSnapshotter(hasher, stringInterner, fileSystem, fileSystemMirror, DirectoryScanner.getDefaultExcludes());
+        return retval;
```

Comment:
I feel this is the wrong place to fix this issue. I think the `CppCompile` task should use the filter pattern for the component when reverting to snapshotting the entire include directory. At this level, we are in the core of Gradle which would add a constraint too hard for everyone using the `FileSystemSnapshotter`.

Reply:
We need a little more information here too.

Is the project directory the source directory, the output directory or both?

Reply:
If you look at the issue, it has nothing to do with CppCompile. The original issue simply copied a README template into the root in a task in a Java project. This just happens to be another use case for fixing this. This particular change addresses the CppCompile issue and the one listed in the issue sited.